app/Rules/UsernameFormatRule.php

```diff
@@ -10,8 +10,9 @@ class UsernameFormatRule implements ValidationRule
 {
     public function validate(string $attribute, mixed $value, Closure $fail): void
     {
+
         if (! is_string($value) || $value !== (new GenerateUsername())($value)) {
-            $fail('The :attribute must be valid');
+            $fail(__('The :attribute must be valid. Only lowercase ASCII characters are allowed. Hyphens can be used. Whitespace, underscores, and multiple hyphens are not permitted. The username cannot start or end with a hyphen.'));
```

Comment:
I think this message makes more sense now :)

Reply:
You can reverse the condition and it will be more readable. Try to do this:

```php
public function validate(string $attribute, mixed $value, Closure $fail): void
{
    $username = (new GenerateUsername())($value);
    
    if (is_string($value) && $value === $username) {
        return;
    }

    $fail(__('The :attribute must be valid. Only lowercase ASCII characters are allowed. Hyphens can be used. Whitespace, underscores, and multiple hyphens are not permitted. The username cannot start or end with a hyphen.'));
}
```

Reply:
So this is what I had initially, but if we look at [GenerateUsername](https://github.com/brendt/rfc-vote/blob/main/app/Actions/GenerateUsername.php#L10C8-L10C8) action we see that it only accepts a User model or a string, if I change the code to what you are proposing there is a chance to have unexpected errors because the type of value in the rule is mixed.

I do agree with you that your suggestion is much cleaner.

We can also write it like this: 
```php
public function validate(string $attribute, mixed $value, Closure $fail): void
{
    if(! is_string($value)) return;

    $username = (new GenerateUsername())($value);
    
    if ($value === $username) {
        return;
    }

    $fail(__('The :attribute must be valid. Only lowercase ASCII characters are allowed. Hyphens can be used. Whitespace, underscores, and multiple hyphens are not permitted. The username cannot start or end with a hyphen.'));
}
```

This is the question, should this custom rule care to validate if the value passed to it is string? Or it should just not care about that, as it is the responsibility of the developer who uses this custom rule to make sure it is used to validate username related fields which are always strings?